airbyte-config/config-models/src/main/java/io/airbyte/config/helpers/StateMessageHelper.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.config.helpers;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JsonNode;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.config.StateType;
+import io.airbyte.config.StateWrapper;
+import io.airbyte.protocol.models.AirbyteStateMessage;
+import io.airbyte.protocol.models.AirbyteStateMessage.AirbyteStateType;
+import java.util.List;
+import java.util.Optional;
+
+public class StateMessageHelper {
+
+  public static class AirbyteStateMessageListTypeReference extends TypeReference<List<AirbyteStateMessage>> {}
+
+  /**
+   * This a takes a json blob state and tries return either a legacy state in the format of a json
+   * object or a state message with the new format which is a list of airbyte state message.
+   *
+   * @param state
+   * @return Either a json blob (on the left) or a structure state message.
+   */
+  public static Optional<StateWrapper> getTypedState(JsonNode state) {
+    if (state == null) {
+      return Optional.empty();
+    } else {
+      try {
+        List<AirbyteStateMessage> stateMessages = Jsons.object(state, new AirbyteStateMessageListTypeReference());
+        if (stateMessages.size() == 1 && stateMessages.get(0).getStateType() == AirbyteStateType.GLOBAL) {
+          return Optional.of(new StateWrapper()
+              .withStateType(StateType.GLOBAL)
+              .withGlobal(stateMessages.get(0)));
+        } else if (stateMessages.size() >= 1
+            && stateMessages.stream().filter(stateMessage -> stateMessage.getStateType() != AirbyteStateType.STREAM).toList().isEmpty()) {
+          return Optional.of(new StateWrapper()
+              .withStateType(StateType.STREAM)
+              .withStateMessages(stateMessages));
+        } else {
+          throw new IllegalStateException("Unexpected state blob");
+        }
+
+      } catch (final IllegalArgumentException e) {
+        return Optional.of(new StateWrapper()
+            .withStateType(StateType.LEGACY)
+            .withLegacyState(state));
+      }
```

Comment:
this is a nit but i would suggest moving this into a helper method. the catch being so far from where the exception is thrown isn't very easy to parse.

```
  /**
   * Parses a state object as a list of AirbyteStateMessages. If cannot be parsed as such, returns an empty optional.
   * @param state - state blob of unknown type either legacy (just a json blob) or global or stream which will be a list of AirbyteStateMessages.
   * @return optional of list of AirbyteStateMessages if parseable, otherwise empty optional.
   */
  private static Optional<List<AirbyteStateMessage>> getStateAsListOrEmpty(JsonNode state)
    try {
      return Optional.of(Jsons.object(state, new AirbyteStateMessageListTypeReference()));
    } catch (final IllegalArgumentException e) {
      return Optional.empty();
    }
  }
```